state/pool.go

```diff
@@ -37,19 +48,78 @@ func (p *StatePool) Get(modelUUID string) (*State, error) {
 	p.mu.Lock()
 	defer p.mu.Unlock()
 
-	st, ok := p.pool[modelUUID]
+	item, ok := p.pool[modelUUID]
+	if ok && item.remove {
+		// We don't want to allow increasing the refcount of a model
+		// that's been removed.
+		return nil, errors.Errorf("model %v has been removed", modelUUID)
+	}
 	if ok {
-		return st, nil
+		item.references++
+		return item.state, nil
 	}
 
 	st, err := p.systemState.ForModel(names.NewModelTag(modelUUID))
 	if err != nil {
 		return nil, errors.Annotatef(err, "failed to create state for model %v", modelUUID)
 	}
-	p.pool[modelUUID] = st
+	p.pool[modelUUID] = &PoolItem{state: st, references: 1}
 	return st, nil
 }
 
+// Put indicates that the client has finished using the State. If the
+// state has been marked for removal, it will be closed and removed
+// when the final Put is done.
+func (p *StatePool) Put(modelUUID string) error {
```

Comment:
Put() seems like a poor choice of name for this? It implies it should be the opposite of Get(), which returns a State.
Could we call it Release() ?